python/lance/types/box.py

```diff
@@ -30,84 +33,211 @@ def __init__(self):
         # return the sliced values properly
         values = self.storage.values
         # If this is a full-length array then just default Arrow
-        if len(self) == len(values) / 4:
+        if len(self) == len(values) / (self.ndims * 2):
             return values
         else:
             # If this is a slice then we need to compute manually
             return pc.list_flatten(self.storage)
 
     def to_numpy(self, zero_copy_only=True):
+        """
+        Return a homogeneous 2-d numpy array where each row represents the
+        2 points defining each bounding box
+
+        For a 2-d box, each row will have (xmin, ymin, xmax, ymax)
+        For a 3-d box, each row will have (xmin, ymin, zmin, xmax, ymax, zmax)
+        """
         return (
             self.flatten()
             .to_numpy(zero_copy_only=zero_copy_only)
-            .reshape((len(self), 4))
+            .reshape((len(self), (self.ndims * 2)))
         )
 
+    def get_length(self, axis: Union[int, str]):
+        axis = self._sanitize_axis(axis)
+        return self.get_max(axis) - self.get_min(axis) + 1
+
+    def get_min(self, axis: Union[int, str]):
+        """
+        Get a 1-d numpy array of the minimums for each box along given axis
+
+        Parameters
+        ----------
+        axis: int or str
+            0 is 'x', 1 is 'y', 2 is 'z`
+        """
+        axis = self._sanitize_axis(axis)
+        return self.to_numpy()[:, axis]
+
+    def get_max(self, axis: Union[int, str]):
+        """
+        Get a 1-d numpy array of the maximums for each box along given axis
+
+        Parameters
+        ----------
+        axis: int or str
+            0 is 'x', 1 is 'y', 2 is 'z`
+        """
+        axis = self._sanitize_axis(axis)
+        return self.to_numpy()[:, self.ndims + axis]
+
+    def _sanitize_axis(self, axis):
+        if isinstance(axis, str):
+            axis = {"x": 0, "y": 1, "z": 2}[axis]
+        if axis < 0 or axis >= self.ndims:
+            raise ValueError(
+                f"Axis {axis} is greater the number of box " f"dimensions in this array"
+            )
+        return axis
+
     @property
     def xmin(self) -> np.ndarray:
         """Return a numpy array of the min X-coord of each box"""
-        return self.to_numpy()[:, 0]
+        return self.get_min("x")
 
     @property
     def ymin(self) -> np.ndarray:
         """Return a numpy array of the min Y-coord of each box"""
-        return self.to_numpy()[:, 1]
+        return self.get_min("y")
 
     @property
     def xmax(self) -> np.ndarray:
         """Return a numpy array of the max X-coord of each box"""
-        return self.to_numpy()[:, 2]
+        return self.get_max("x")
 
     @property
     def ymax(self) -> np.ndarray:
         """Return a numpy array of the max Y-coord of each box"""
-        return self.to_numpy()[:, 3]
+        return self.get_max("y")
+
+    def _box_sizes(self):
+        sizes = self.get_length(0)
```

Comment:
shall we do 

```
sizes = np.ones_like(self.dims[0])
for i in range(self.ndims):
    sizes *= self.get_length(i)
```

Or prob a `reduce` .

Reply:
wanted to save the allocation and one array mult operation. i guess reduce could work well.

Reply:
changed to using `np.multiply.reduce`